### PR TITLE
fix: abort in-flight snapshots on Close

### LINKED ISSUE #7252: GoLevelDB closing panic
while running a node, I consistenly see:
```
2:42PM INF service stop impl={"Data":{},"Logger":{"Logger":{}}} module=p2p msg={} peer={"id":"c7b0ad37748a04bf73a04824e57977f7d53dd4d1","ip":"61.111.3.70","port":20470}                        14:42:27 [0/1854]
2:42PM ERR error while stopping peer error="already stopped" module=p2p
2:42PM INF service stop impl={"Logger":{"Logger":{}}} module=p2p msg={} peer={"id":"0bba2f23a43ca33911cd50c2fc48c5727bc83af5","ip":"61.111.3.52","port":20470}
2:42PM ERR Stopping peer for error err="read tcp 147.182.189.182:32846->61.111.3.52:20470: use of closed network connection" module=p2p peer={"Data":{},"Logger":{"Logger":{}}} reactor=p2p
2:42PM ERR error while stopping peer error="already stopped" module=p2p
2:42PM INF service stop impl={"Data":{},"Logger":{"Logger":{}}} module=p2p msg={} peer={"id":"0bba2f23a43ca33911cd50c2fc48c5727bc83af5","ip":"61.111.3.52","port":20470}
2:42PM INF service stop impl={"Logger":{"Logger":{}}} module=p2p msg={} peer={"id":"3b5781469fb0014f9293833cfcd0de13be188fbd","ip":"180.189.55.7","port":26656}
2:42PM INF service stop impl={"Data":{},"Logger":{"Logger":{}}} module=p2p msg={} peer={"id":"3b5781469fb0014f9293833cfcd0de13be188fbd","ip":"180.189.55.7","port":26656}
2:42PM INF service stop impl={"Logger":{"Logger":{}}} module=p2p msg={} peer={"id":"940dfbbb0884185984b098457a21e523391545ac","ip":"95.216.78.225","port":12056}
2:42PM INF service stop impl={"Data":{},"Logger":{"Logger":{}}} module=p2p msg={} peer={"id":"940dfbbb0884185984b098457a21e523391545ac","ip":"95.216.78.225","port":12056}
2:42PM INF service stop impl={"Logger":{"Logger":{}},"Switch":{"Logger":{"Logger":{}}}} module=server msg={}
2:42PM INF service stop impl={"Logger":{"Logger":{}},"Switch":{"Logger":{"Logger":{}}}} module=blocksync msg={}                     
2:42PM INF service stop impl={"Logger":{"Logger":{}}} module=blocksync msg={}
2:42PM INF finalized block block_app_hash=1F1C3291E0D61DCF698D3124CFCD04CDA87BD3B71580CC541F858FB2F69131F9 height=8485788 module=state num_txs_res=3 num_val_updates=0
2:42PM INF service stop impl={"Logger":{"Logger":{}},"Metrics":{"ApplicationRejectedProposals":{},"BlockGossipPartsReceived":{},"BlockIntervalSeconds":{},"BlockParts":{},"BlockSizeBytes":{},"BlockTimeSeconds":
{},"ByzantineValidators":{},"ByzantineValidatorsPower":{},"ChainSizeBytes":{},"CommittedHeight":{},"DuplicateBlockPart":{},"DuplicateVote":{},"FullPrevoteDelay":{},"Height":{},"LateVotes":{},"MissingValidators
":{},"MissingValidatorsPower":{},"NumTxs":{},"ProposalCreateCount":{},"ProposalReceiveCount":{},"ProposerMissedProposals":{},"QuorumPrevoteDelay":{},"RoundDurationSeconds":{},"RoundVotingPowerPercent":{},"Roun
ds":{},"StartHeight":{},"StepDurationSeconds":{},"TimedOutProposals":{},"TotalTxs":{},"ValidatorLastSignedHeight":{},"ValidatorMissedBlocks":{},"ValidatorPower":{},"Validators":{},"ValidatorsPower":{},"VoteExt
ensionReceiveCount":{}},"Switch":{"Logger":{"Logger":{}}}} module=consensus msg={}
2:42PM ERR Not stopping State service -- has not been started yet impl={"Logger":{"Logger":{}},"StartedPrecommitSleep":{}} module=consensus
2:42PM ERR Error stopping consensus state err="not started" module=consensus
2:42PM INF service stop impl={"Logger":{"Logger":{}},"Switch":{"Logger":{"Logger":{}}}} module=evidence msg={}
2:42PM INF service stop impl={"Logger":{"Logger":{}},"Switch":{"Logger":{"Logger":{}}}} module=statesync msg={}
2:42PM INF service stop impl={"BlockMaxBytes":8388608,"Logger":{"Logger":{}},"Switch":{"Logger":{"Logger":{}}}} module=propagation msg={}
2:42PM INF service stop impl={"Logger":{"Logger":{}},"Switch":{"Logger":{"Logger":{}}}} module=pex msg={}
2:42PM INF service stop book=/root/.celestia-app/config/addrbook.json impl={"Logger":{"Logger":{}}} module=p2p msg={}
2:42PM INF Closing rpc listener listener={"Listener":{}} module=server
2:42PM INF RPC HTTP server stopped err="accept tcp 127.0.0.1:26657: use of closed network connection" module=rpc-server
2:42PM ERR Error serving server err="accept tcp 127.0.0.1:26657: use of closed network connection" module=server
2:42PM INF Closing rpc listener listener={"Listener":{}} module=server
2:42PM ERR Stopped accept routine, as transport is closed module=p2p numPeers=0
2:42PM INF Closing blockstore module=server
2:42PM INF Closing statestore module=server
2:42PM INF Closing evidencestore module=server
2:42PM INF stopping embedded app for version active_app_version=5 module=server
2:42PM INF caught signal module=server signal=interrupt
2:42PM INF stopping the ABCI server... module=server
2:42PM INF service stop impl=ABCIServer module=abci-server msg="Stopping ABCIServer service"
2:42PM INF Closing application.db module=server
2:42PM ERR snapshot failed; exporter error err="iavl export failed for version 8463000: leveldb: closed" module=server store=icahost
panic: leveldb: closed

goroutine 911 [running]:
cosmossdk.io/store/pruning.(*Manager).HandleSnapshotHeight(0xc000203880, 0x812298)
        cosmossdk.io/store@v1.1.2/pruning/manager.go:93 +0x36a
cosmossdk.io/store/rootmulti.(*Store).PruneSnapshotHeight(0x1b7f705?, 0xc002259b80?)
        cosmossdk.io/store@v1.1.2/rootmulti/store.go:357 +0x17
cosmossdk.io/store/snapshots.(*Manager).Create(0xc002259b80, 0x812298)
        cosmossdk.io/store@v1.1.2/snapshots/manager.go:188 +0x282
cosmossdk.io/store/snapshots.(*Manager).snapshot(0xc002259b80, 0x812298)
        cosmossdk.io/store@v1.1.2/snapshots/manager.go:531 +0x15a
created by cosmossdk.io/store/snapshots.(*Manager).SnapshotIfApplicable in goroutine 910
        cosmossdk.io/store@v1.1.2/snapshots/manager.go:515 +0x85
2026/05/12 14:42:28 Process finished with error: exit status 2
2:42PM INF stopping gRPC connection for ABCI module=server
```

whenever I stop it.

We need to investigate the root cause and fix it.

### PR DESCRIPTION
**Pull request by NikhilSharmaWe**
## Summary
- Fixes [celestia-app#7252](https://github.com/celestiaorg/celestia-app/issues/7252): shutdown panic (`leveldb: closed` / `pebble: closed`) when a state-sync snapshot is in flight.
- Root cause: `SnapshotIfApplicable` runs async; `BaseApp.Close` closed `application.db` before aborting that work; `HandleSnapshotHeight` panics on `SetSync` errors.
- Changes:
  - Close snapshot manager **before** `application.db`
  - `Manager.Close` aborts in-flight export/restore and waits for unwind (does not wait for a full snapshot to finish)
  - Skip `PruneSnapshotHeight` while closing; log DB write errors in `HandleSnapshotHeight` instead of panicking
  - Make `ChunkWriter` safe for concurrent `Write` vs `CloseWithError`; abort promptly via `abortAwareWriter`

## Test plan
- [x] `go test ./baseapp/ -run TestBaseApp_Close_`
- [x] `go test ./store/snapshots/ ./store/pruning/ ./store/rootmulti/ -run 'Close|HandleSnapshotHeight_DbErr|ManagerCloseAborts'`
- [x] Local A/B against celestia-app with `replace` to this branch vs published `v0.52.8` / `store@v1.1.3-celestia.1`: unfixed reproduces `panic: leveldb/pebble: closed` via `HandleSnapshotHeight`; this branch does not
## Notes
Matches intended shutdown behavior: cancel in-flight snapshot work on stop, without waiting for snapshot creation to complete.

### CODE REVIEW

**Comment by NikhilSharmaWe** [comment]
cc @rach-id

**Review by rach-id — COMMENTED** [review]
Thanks for working on this 🙏 

I think the design could be simplified. Right now, it's hard to review and will be hard to maintain in the future.  Here are couple propositions to make this simpler:

<details>
  <summary>Alternative 1: </summary>

```
## A substantially simpler design

  I would narrow this PR to snapshot creation and use one lifecycle mechanism:

  1. Add closed, activeWriter, and one manager-level completion primitive—either a WaitGroup or operation done channel.
  2. Register an operation synchronously in begin, under the same mutex that checks closed.
  3. Deregister it in end.
  4. In Close:
      - Set closed.
      - Close the active writer with ErrAborted.
      - Wait for registered work to unwind.
      - Close the snapshot metadata DB.

  5. Then let BaseApp.Close close application.db.

  This removes:

  - sync.Cond and both wait loops.
  - The local snapshotWG in every Create.
  - The private snapshotAbortAware hook.
  - The pruning error-policy change.
  - Most restore-specific shutdown handling.

  The abortAwareWriter can remain because it makes ordinary rootmulti cancellation prompt between snapshot items. The concurrent ChunkWriter work is also partly necessary, although its implementation can be
  tightened—closeChOnce, for example, is redundant when closed and channel closure are already protected by the same mutex.

  I expect the manager lifecycle portion could be reduced by roughly half without giving up the behavior needed for the reported issue.
```
</details>

<details>
  <summary>Alternative 2: </summary

```
What's necessary (keep as-is)

- baseapp.go reorder — trivial and correct. (The strconv/sync import reorder is an unrelated drive-by, but harmless.)
- HandleSnapshotHeight panic → log — correct; a best-effort persistence write should never crash the process.
- closed flag + beginLocked rejection + SnapshotIfApplicable guard + ErrAborted — the minimal state needed.
- abortAwareWriter — this is actually the workhorse of the whole PR. The write path is proto → zlib → 10MB bufio → ChunkWriter, so an abort signaled only at the ChunkWriter level wouldn't be observed until ~10MB of compressed output flushes. Checking abortCh on every WriteMsg is the right place.
- Skipping PruneSnapshotHeight when closing — needed; that's the exact call that panicked.
- The restore-abort path in Close — shutdown during state-sync restore is plausible, and the chRestore nil-ing to avoid double-close looks correct.

Where it can be simplified a lot

1. The concurrent-abort machinery is mostly redundant with guarantees that already exist. This is the big one. The PR builds two heavy mechanisms so that Manager.Close can forcibly unblock an export goroutine stuck mid-write:

- The full ChunkWriter thread-safety rewrite (~100 lines: chunkLocked with its unlock/relock dance around pipe.Close, prepareWrite, addWritten, closeChOnce) — the subtlest, most bug-prone code in the diff.
- The activeWriter registration + idle cond var + two-phase wait loop in Manager.Close (~50 lines), which exists solely so Close can hunt down the writer and call CloseWithError on it from another goroutine.

But trace what happens with only abortAwareWriter + a plain "wait until the operation ends": the base code already has defer DrainChunks(chunks) in Store.Save, and Save runs synchronously inside Create — so the chunk channel always has an active consumer until the writer itself closes it. A blocked pipe.Write or ch <- pr send therefore always unblocks within one chunk's consumption, at which point the next WriteMsg hits the closed abortCh and the export unwinds through the existing single-goroutine CloseWithError path. No cross-goroutine CloseWithError ever happens, so ChunkWriter never needs to be thread-safe, and Close reduces to: set closed, close abortCh, handle restore, then wait on a cond/WaitGroup for operation == opNone. That deletes ~150 lines and, more importantly, deletes the code a reviewer can't easily prove correct.

The only thing lost is instantaneous unblocking of a write that's mid-pipe — the simpler design's abort latency is bounded by writing one in-flight chunk to disk, which is fine for shutdown. If the authors do want immediate unblocking, there's still a simpler way than writer-hunting from Close: a small watcher goroutine inside createSnapshot (select { case <-abortCh: streamWriter.CloseWithError(ErrAborted); case <-done: }). That keeps CloseWithError racing Write (so some ChunkWriter locking survives), but it eliminates activeWriter, the cond var, and the two-phase loop entirely, because the goroutine that owns the writer is the one aborting it — no registration race exists by construction.

2. snapshotAbortAware / SetSnapshotAbortCh is test scaffolding leaked into production API. No production snapshotter implements it — its only implementer is hungSnapshotter in helpers_test.go, which needs it because it blocks without ever writing (so abortAwareWriter can't reach it). The PR already contains the better test double: writingSnapshotter, which exercises the real abort path. Drop the interface and the type-assertion injection in NewManager, and rework the hung-snapshotter tests to close the snapshotter before manager.Close (or use the writing variant).

3. Minor trims:
- The outer for loop in Manager.Close can't actually iterate more than once meaningfully — beginLocked rejects new operations once closed is set, so at most one writer ever registers. A single pass (wait-for-registration-or-idle → abort → wait-for-idle) suffices.
- Create returns ErrAborted even when Save completed successfully and a valid snapshot was persisted (the if m.isClosed() check after snapshotWG.Wait()). Prefer return snapshot, err when err == nil; only map to ErrAborted when there was an actual failure.
- chunkLocked's unlock/relock around pipe.Close() is unnecessary even within the current design — io.PipeWriter.Close never blocks — and it's what forces the awkward double closed recheck. (Moot if suggestion 1 is taken.)

```
</details>

Read these alternatives, choose the better one and feed it to your LLM and it should give you better results 🙏

**Comment by NikhilSharmaWe** [comment]
@rach-id Thanks for the review.

Originally, I tried to cover every shutdown edge case I could think of, including an export stuck mid-write, and a test snapshotter that blocks without writing. Therefore, the PR ended up with activeWriter, the cond wait loops, the ChunkWriter concurrency changes, and SetSnapshotAbortCh. It stops the panic, but it’s more complex than it needed to be.

Alternative 2 makes more sense for this. What’s actually needed is:
- close the snapshot manager before application.db
- signal abort so abortAwareWriter fails on the next WriteMsg
- skip PruneSnapshotHeight while closing / don’t panic in HandleSnapshotHeight
- wait for the in-flight Create/Restore to finish unwinding

Save already reads the chunk channel, so Close does not need to find the writer and call CloseWithError on it. On shutdown, it is fine if we wait until the current chunk is consumed and the next WriteMsg sees the abort.

Will simplify the PR that way: keep abortAwareWriter and the pruning/restore handling, remove the writer-registration machinery and ChunkWriter rewrite, and stop returning ErrAborted when Save already succeeded.

Will push an update soon.

**Comment by NikhilSharmaWe** [comment]
Pushed the simplification.

Dropped the concurrent ChunkWriter changes, activeWriter / CloseWithError-from-Close, and SetSnapshotAbortCh. Close now just marks closed, closes abortCh, handles restore teardown, waits for the in-flight op to finish, then closes the snapshot DB. abortAwareWriter still does the actual cancel on the next WriteMsg.

Also stopped mapping a successful Save to ErrAborted, and updated the tests to use writingSnapshotter for the abort case.

Local smoke still looks good: unfixed panics in HandleSnapshotHeight, this branch does not.

**Comment by rach-id** [comment]
@claude review once

File: store/snapshots/manager.go [hunk 1]
```diff
@@ -552,7 +622,49 @@ func (m *Manager) snapshot(height int64) {
...
+		<-chRestoreDone
+		m.mtx.Lock()
+		if m.operation == opRestore {
+			m.endLocked()
+		}
+	}
+
+	for m.operation != opNone {
+		m.idle.Wait()
+	}
```
Comment (rach-id) [inline store/snapshots/manager.go:698]: I'm skeptical of this code block. It's complicated to fully understand. Could you try and simplify it a bit?

**Review by rach-id — COMMENTED** [review]
Thanks for the simplification. Now it looks better. However, given this is critical code, we can't merge it until thoroughly reviewed. So, I'll do a few review rounds before approving it.

Some LLM review actionable feedback:
```
  1. [P1] Close can still wait forever—or for the full snapshot.
     Close waits unconditionally for operation == opNone (https://github.com/celestiaorg/cosmos-sdk/blob/1fa81b698bb4520500ff6d975b2a5d045dbb1830/store/snapshots/manager.go#L664-L665), but cancellation is only
     observed by the next WriteMsg. A snapshotter or extension blocked before writing never observes it. The PR’s own modified helper explicitly warns about this
     (https://github.com/celestiaorg/cosmos-sdk/blob/1fa81b698bb4520500ff6d975b2a5d045dbb1830/store/snapshots/helpers_test.go#L207-L211), and the previous busy-manager Close test was removed. I reproduced a
     permanent shutdown hang with a valid Snapshotter blocked before its first write. This contradicts the linked issue’s requirement to stop gracefully without waiting for snapshot completion. Cancellation
     must reach the snapshotter/extension operation itself, with a regression test covering pre-write blocking.

  2. [P1] A restore setup error permanently deadlocks Close.
     Restore sets operation = opRestore before creating the snapshot directory (https://github.com/celestiaorg/cosmos-sdk/blob/1fa81b698bb4520500ff6d975b2a5d045dbb1830/store/snapshots/manager.go#L346-L357). If
     MkdirAll fails, it returns without calling endLocked and without initializing chRestoreDone. The new Close therefore has nothing to abort and waits forever
     (https://github.com/celestiaorg/cosmos-sdk/blob/1fa81b698bb4520500ff6d975b2a5d045dbb1830/store/snapshots/manager.go#L655-L665). I reproduced this by placing a file where the restore directory must be
     created. Roll back the operation on every post-beginLocked error, or complete setup before beginning the operation.

  3. [P2] Concurrent Close callers do not share completion.
     Once the first caller sets closed, a second caller immediately returns nil (https://github.com/celestiaorg/cosmos-sdk/blob/1fa81b698bb4520500ff6d975b2a5d045dbb1830/store/snapshots/manager.go#L636-L640),
     even while the first is waiting for snapshot work and has not closed the metadata DB. With concurrent BaseApp.Close calls, the second caller can proceed to close application.db, recreating the exact
     read-after-close race this PR addresses. All callers should wait on a shared close-completion channel and receive the same result.

4. The HandleSnapshotHeight panic→log change is broader than the shutdown fix. A genuine SetSync failure during normal operation (disk full, corruption) is now just an error log. Consequence is mild — the heights stay in memory so pruning proceeds this run; they're only lost if the node crashes before pruning, leaving those heights permanently unpruned. Similarly, skipping PruneSnapshotHeight on abort means an aborted snapshot's height is never registered for pruning — a small, permanent per-abort disk leak. Both are reasonable tradeoffs versus crashing the node from a background goroutine, but the changelog entry could mention the durability implication.

Note: upstream cosmos-sdk (checked release/v0.53.x) still closes application.db before the snapshot manager, so this bug class exists upstream too — this fix (or at least the reordering) is a good candidate to upstream.
```

File: store/snapshots/manager.go (hunk 1 above)
Comment (NikhilSharmaWe) [reply store/snapshots/manager.go:698]: Separated Close into clearer steps. The restore wait is still needed otherwise `opRestore` never gets cleared and `Close` can hang.

**Comment by NikhilSharmaWe** [comment]
Addressed the points your review:

1. Pre-write abort: `rootmulti` observes `abortCh` via `SetSnapshotAbortCh`, with a regression test.
2. Restore `MkdirAll` failure: rolls back `opRestore` so `Close` doesn’t hang.
3. Concurrent `Close`: callers share completion.
4. Changelog: notes the prune/durability tradeoff.

Also simplified the `Close` block a bit for readability.

`Close` still waits until in-flight `Snapshot()` returns after abort (needed so we don’t close `application.db` under a live export).

**Comment by rach-id** [comment]
I'll make this PR draft until I get time to review it thoroughly later. I have other priorities at the moment. Thanks for your contributions 🙏

**Comment by NikhilSharmaWe** [comment]
Commenting on removing stale
